### PR TITLE
add new register function for adding segment.io id

### PR DESCRIPTION
**Pull request by Xingyuj**
add new register function for adding segment.io id IOS-1264

### CODE REVIEW

File: Example_DynamicFramework/Demo/AppDelegate.m [hunk 1]
```diff
@@ -18,7 +18,7 @@
 #import "AppDelegate.h"
 #import "SampleCaseViewController.h"
 #import "AppKeyChoiceViewController.h"
-
+#import <Analytics/SEGAnalytics.h>
```
Comment (ChristineXYS) [inline Example_DynamicFramework/Demo/AppDelegate.m]: In this case, streethawk must depends on segment to work. You need to add this into pod's dependency. Now it depends on customer to add "pod 'Analytics', '~> 3.0'".

File: Example_DynamicFramework/Demo/AppDelegate.m [hunk 2]
```diff
@@ -76,6 +76,16 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     //    StreetHawk.notificationTypes = UIRemoteNotificationTypeAlert;
     //    StreetHawk.isDefaultLocationServiceEnabled = NO;
     [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(installRegisterSuccessHandler:) name:SHInstallRegistrationSuccessNotification object:nil];
+    
+    SEGAnalyticsConfiguration *configuration = [SEGAnalyticsConfiguration configurationWithWriteKey:@"Z3KDihiA42uXBPr82xULjKA6Fb0p0v38"];
```
Comment (ChristineXYS) [inline Example_DynamicFramework/Demo/AppDelegate.m]: Where is this key from? Customer's register key in segment?

File: Example_DynamicFramework/Demo/AppDelegate.m [hunk 3]
```diff
@@ -76,6 +76,16 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
...
+    configuration.trackApplicationLifecycleEvents = YES; // Enable this to record certain application events automatically!
+    configuration.recordScreenViews = YES; // Enable this to record screen views automatically!
+    [SEGAnalytics setupWithConfiguration:configuration];
+    
+    
+    StreetHawk.autoIntegrateAppDelegate = YES;
+    [StreetHawk registerInstallForApp:appKey segmentID:[[SEGAnalytics sharedAnalytics] getAnonymousId] withDebugMode:YES];
```
Comment (ChristineXYS) [inline Example_DynamicFramework/Demo/AppDelegate.m]: Does getAnonymousId change in each call?

File: StreetHawk/Classes/Core/Publish/SHApp.m [hunk 4]
```diff
@@ -34,7 +34,7 @@
 
 #define APPKEY_KEY                          @"APPKEY_KEY" //key for store "app key", next time if try to read appKey before register, read from this one.
 #define INSTALL_SUID_KEY                    @"INSTALL_SUID_KEY"
-
+#define SEGMENT_ID                          @"SEGMENT_ID" //segment.io id
```
Comment (ChristineXYS) [inline StreetHawk/Classes/Core/Publish/SHApp.m]: What's this for?

File: StreetHawk/Classes/Core/Publish/SHApp.m [hunk 5]
```diff
@@ -1905,6 +1917,7 @@ -(void)registerOrUpdateInstallWithHandler:(SHCallbackHandler)handler
 -(BOOL)checkInstallChangeForLaunch
 {
     NSString *sentAppKey = [[NSUserDefaults standardUserDefaults] objectForKey:SentInstall_AppKey];
+    NSString *sentSegmentId = [[NSUserDefaults standardUserDefaults] objectForKey:SentInstall_SegmentId];
```
Comment (ChristineXYS) [inline StreetHawk/Classes/Core/Publish/SHApp.m]: In later code it should check whether this change, and do something.

File: Example_DynamicFramework/Demo/AppDelegate.m (hunk 1 above)
Comment (ChristineXYS) [reply Example_DynamicFramework/Demo/AppDelegate.m]: More importantly, this file need to be added into Pointzi built output library. Check how AFNetworking does.

Comment (ChristineXYS) [reply Example_DynamicFramework/Demo/AppDelegate.m]: I think carefully, it's too heavy to add additional dependency only for `[[SEGAnalytics sharedAnalytics] getAnonymousId]`. We can use `performSelector` to archive this.

Comment (asyncmind0) [reply Example_DynamicFramework/Demo/AppDelegate.m]: I think @Xingyuj commited his experimental code ? this is not required

File: Example_DynamicFramework/Demo/AppDelegate.m (hunk 2 above)
Comment (asyncmind0) [reply Example_DynamicFramework/Demo/AppDelegate.m]: obviously this should be in the sample app codebase not in the library codebase

File: Example_DynamicFramework/Demo/AppDelegate.m (hunk 3 above)
Comment (asyncmind0) [reply Example_DynamicFramework/Demo/AppDelegate.m]: segmentId  should be just a string, anonymousid may or may not change, better to handle both cases.

But you need to make sure that you get anonymous id after segment integration has been initialized.

https://segment.com/docs/sources/mobile/ios/#how-do-i-know-when-a-destination-is-initialized-

you have to initialize pointzi in the callback to make sure that you get a segment anonymousId when pointzi starts

**Review by asyncmind0 — CHANGES_REQUESTED** [review]
Looks like there is still lot of work to be done ?

File: Example_DynamicFramework/Demo/AppDelegate.m (hunk 1 above)
Comment (Xingyuj) [reply Example_DynamicFramework/Demo/AppDelegate.m]: Alright, I thought I should add the segment for our demo app as well, I will get rid of those changes

File: StreetHawk/Classes/Core/Publish/SHApp.m (hunk 5 above)
Comment (Xingyuj) [reply StreetHawk/Classes/Core/Publish/SHApp.m]: So, if segment id won't be changed in the lifecycle of an install, we don't need to put it here right? I will delete this line

File: Example_DynamicFramework/Demo/AppDelegate.m [hunk 6]
```diff
@@ -18,8 +18,9 @@
 #import "AppDelegate.h"
 #import "SampleCaseViewController.h"
 #import "AppKeyChoiceViewController.h"
-
 #import <StreetHawkCore/StreetHawkCore.h>
+#import <Analytics/SEGAnalytics.h>
+#import <Analytics/SEGIntegrationsManager.h>
```
Comment (ChristineXYS) [inline Example_DynamicFramework/Demo/AppDelegate.m:23]: As the previous comment said, you cannot add segment.io in this way without handling mandatory dependency.

Comment (asyncmind0) [reply Example_DynamicFramework/Demo/AppDelegate.m:23]: this is the demo app, so can we add segment as a dependency ?

**Review by ChristineXYS — APPROVED** [review]

**Review by asyncmind0 — APPROVED** [review]

**Pull request merged**